### PR TITLE
zig: assert there is an exception when .zero is returned

### PR DESCRIPTION
**Pull request by nektro**
cc https://github.com/oven-sh/bun/issues/15141

adds some assertions and fixes the trips

### CODE REVIEW

**Comment by robobun** [comment]
@nektro, your commit 9be11afa76712485fd7171989917b35ab2e5266b is building: [`#6765`](https://buildkite.com/bun/bun/builds/6765)

File: src/bun.js/bindings/bindings.zig [hunk 1]
```diff
@@ -6780,6 +6788,14 @@ pub fn toJSHostFunction(comptime Function: JSHostZigFunction) JSC.JSHostFunction
 
 // XXX: temporary
 pub fn toJSHostValue(globalThis: *JSGlobalObject, value: error{ OutOfMemory, JSError }!JSValue) JSValue {
+    if (bun.Environment.allow_assert) {
+        const normal = value catch |err| switch (err) {
+            error.JSError => .zero,
+            error.OutOfMemory => globalThis.throwOutOfMemoryValue(),
+        };
+        bun.assert((normal == .zero) == globalThis.hasException());
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.zig:6797]:
```suggestion
        bun.assert((normal == .zero or normal == .undefined) == globalThis.hasException());
```

File: src/bun.js/bindings/bindings.zig [hunk 2]
```diff
@@ -6770,6 +6770,14 @@ pub fn toJSHostFunction(comptime Function: JSHostZigFunction) JSC.JSHostFunction
             globalThis: *JSC.JSGlobalObject,
             callframe: *JSC.CallFrame,
         ) callconv(JSC.conv) JSC.JSValue {
+            if (bun.Environment.allow_assert) {
+                const value = Function(globalThis, callframe) catch |err| switch (err) {
+                    error.JSError => .zero,
+                    error.OutOfMemory => globalThis.throwOutOfMemoryValue(),
+                };
+                bun.assert((value == .zero) == globalThis.hasException());
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.zig:6780]:
```suggestion
                bun.assert((value == .zero or value == .undefined) == globalThis.hasException());
```

File: src/codegen/generate-classes.ts [hunk 3]
```diff
@@ -1860,10 +1854,7 @@ const JavaScriptCoreBindings = struct {
           output += `
         pub fn ${names.fn}(globalObject: *JSC.JSGlobalObject, callFrame: *JSC.CallFrame) callconv(JSC.conv) JSC.JSValue {
           if (comptime Environment.enable_logs) JSC.markBinding(@src());
-          return @call(.always_inline, ${typeName}.${fn}, .{globalObject, callFrame}) catch |err| switch (err) {
-            error.JSError => .zero,
-            error.OutOfMemory => globalObject.throwOutOfMemoryValue(),
-          };
+          return JSC.toJSHostFunction(${typeName}.${fn})(globalObject, callFrame);
```
Comment (Jarred-Sumner) [inline src/codegen/generate-classes.ts]: The inline was intentional. We cannot add extra functions to the call stack. Please revert this

File: src/codegen/generate-classes.ts [hunk 4]
```diff
@@ -1810,10 +1807,7 @@ const JavaScriptCoreBindings = struct {
           output += `
         pub fn ${names.fn}(thisValue: *${typeName}, globalObject: *JSC.JSGlobalObject, callFrame: *JSC.CallFrame${proto[name].passThis ? ", js_this_value: JSC.JSValue" : ""}) callconv(JSC.conv) JSC.JSValue {
           if (comptime Environment.enable_logs) zig("<d>${typeName}.<r>${name}<d>({})<r>", .{callFrame});
-          return @call(.always_inline, ${typeName}.${fn}, .{thisValue, globalObject, callFrame${proto[name].passThis ? ", js_this_value" : ""}}) catch |err| switch (err) {
-            error.JSError => .zero,
-            error.OutOfMemory => globalObject.throwOutOfMemoryValue(),
-          };
+          return JSC.toJSHostValue(globalObject, ${typeName}.${fn}(thisValue, globalObject, callFrame${proto[name].passThis ? ", js_this_value" : ""}));
```
Comment (Jarred-Sumner) [inline src/codegen/generate-classes.ts]: The inline was intentional. We cannot add extra functions to the call stack. Please revert this

File: src/codegen/generate-classes.ts [hunk 5]
```diff
@@ -1754,10 +1754,7 @@ const JavaScriptCoreBindings = struct {
       output += `
         pub fn ${classSymbolName(typeName, "call")}(globalObject: *JSC.JSGlobalObject, callFrame: *JSC.CallFrame) callconv(JSC.conv) JSC.JSValue {
           if (comptime Environment.enable_logs) zig("${typeName}<d>({})<r>", .{callFrame});
-          return @call(.always_inline, ${typeName}.call, .{globalObject, callFrame}) catch |err| switch (err) {
-            error.JSError => .zero,
-            error.OutOfMemory => globalObject.throwOutOfMemoryValue(),
-          };
+          return JSC.toJSHostFunction(${typeName}.call)(globalObject, callFrame);
```
Comment (Jarred-Sumner) [inline src/codegen/generate-classes.ts]: Here as well

File: src/codegen/generate-js2native.ts [hunk 6]
```diff
@@ -216,10 +216,7 @@ export function getJS2NativeZig(gs2NativeZigPath: string) {
         })}(global: *JSC.JSGlobalObject, call_frame: *JSC.CallFrame) callconv(JSC.conv) JSC.JSValue {`,
         `
           const function = @import(${JSON.stringify(path.relative(path.dirname(gs2NativeZigPath), x.filename))});
-          return @call(.always_inline, function.${x.symbol_target}, .{global, call_frame}) catch |err| switch (err) {
-              error.JSError => .zero,
-              error.OutOfMemory => global.throwOutOfMemoryValue(),
-          };`,
+          return JSC.toJSHostFunction(function.${x.symbol_target})(global, call_frame);`,
```
Comment (Jarred-Sumner) [inline src/codegen/generate-js2native.ts]: Also here

File: src/bun.js/bindings/bindings.zig [hunk 7]
```diff
@@ -6780,6 +6790,14 @@ pub fn toJSHostFunction(comptime Function: JSHostZigFunction) JSC.JSHostFunction
 
 // XXX: temporary
 pub fn toJSHostValue(globalThis: *JSGlobalObject, value: error{ OutOfMemory, JSError }!JSValue) JSValue {
+    if (bun.Environment.allow_assert) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.zig]: Let's avoid making function calls slower on Windows

```suggestion
    if (bun.Environment.isDebug) {
```

File: src/bun.js/bindings/bindings.zig [hunk 8]
```diff
@@ -6770,6 +6772,14 @@ pub fn toJSHostFunction(comptime Function: JSHostZigFunction) JSC.JSHostFunction
             globalThis: *JSC.JSGlobalObject,
             callframe: *JSC.CallFrame,
         ) callconv(JSC.conv) JSC.JSValue {
+            if (bun.Environment.allow_assert) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.zig]: Let's avoid making function calls slower on Windows

```suggestion
    if (bun.Environment.isDebug) {
```

**Pull request merged**